File: Classes/AsynchronousImageView.m
//
//  AsynchronousImageView.m
//  GLOSS
//
//  Created by Слава on 22.10.09.
//  Copyright 2009 Slava Bushtruk. All rights reserved.
//  ---------------------------------------------------
//
//  Modified by Ben Baron for the iSub project.
//

#import "AsynchronousImageView.h"
#import "AsynchronousImageViewDelegate.h"
#import "Imports.h"

//LOG_LEVEL_ISUB_DEBUG

@interface AsynchronousImageView ()
{
    __strong NSString *_coverArtId;
}
@end

@implementation AsynchronousImageView

- (id)initWithFrame:(CGRect)frame coverArtId:(NSString *)artId isLarge:(BOOL)large delegate:(NSObject<AsynchronousImageViewDelegate> *)theDelegate
{
	if ((self = [super initWithFrame:frame]))
	{
		_isLarge = large;
        [self setCoverArtId:artId];
		_delegate = theDelegate;
	}
	return self;
}

- (NSString *)coverArtId
{
	return _coverArtId;
}

- (void)setCoverArtId:(NSString *)artId
{
    // Make sure old activity indicator is gone
    [self.activityIndicator removeFromSuperview];
    self.activityIndicator = nil;
    
    if (self.coverArtDAO)
    {
        [self.coverArtDAO cancelLoad];
        self.coverArtDAO.delegate = nil;
        self.coverArtDAO = nil;
    }
    
    _coverArtId = [artId copy];
    self.image = self.coverArtDAO.defaultCoverArtImage;
    
    if (artId)
    {
        self.coverArtDAO = [[SUSCoverArtDAO alloc] initWithDelegate:self coverArtId:self.coverArtId isLarge:self.isLarge];
        if (self.coverArtDAO.isCoverArtCached)
        {
            self.image = self.coverArtDAO.coverArtImage;
        }
        else
        {
            if (_coverArtId && self.isLarge)
            {
                self.activityIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
                self.activityIndicator.autoresizingMask = UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin | UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleBottomMargin;
                self.activityIndicator.center = CGPointMake(self.width/2, self.height/2);
                [self addSubview:self.activityIndicator];
                [self.activityIndicator startAnimating];
            }
            [self.coverArtDAO startLoad];
        }
    }
}

#pragma mark -
#pragma mark Handle User Input

- (void)reloadCoverArt
{
	/*if(musicS.currentSongObject.coverArtId)
	{
		musicS.coverArtUrl = nil;
		if (SCREEN_SCALE() == 2.0)
		{
			musicS.coverArtUrl = [[NSURL alloc] initWithString:[NSString stringWithFormat:@"%@%@&size=640", [appDelegateS getBaseUrl:@"getCoverArt.view"], musicS.currentSongObject.coverArtId]];
		}
		else
		{	
			musicS.coverArtUrl = [[NSURL alloc] initWithString:[NSString stringWithFormat:@"%@%@&size=320", [appDelegateS getBaseUrl:@"getCoverArt.view"], musicS.currentSongObject.coverArtId]];
		}
		[self loadImageFromURLString:[musicS.coverArtUrl absoluteString]];
	}
	else 
	{
		self.image = [UIImage imageNamed:@"default-album-art"];
	}*/
}

-(void)oneTap
{
//DLog(@"Single tap");
}

-(void)twoTaps
{
//DLog(@"Double tap");
	//[self reloadCoverArt];
}

-(void)threeTaps
{
//DLog(@"Triple tap");
}

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event 
{
	// Detect touch anywhere
	UITouch *touch = [touches anyObject];
	
	switch ([touch tapCount]) 
	{
		case 1:
			[self performSelector:@selector(oneTap) withObject:nil afterDelay:.5];
			break;
			
		case 2:
			[NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(oneTap) object:nil];
			[self performSelector:@selector(twoTaps) withObject:nil afterDelay:.5];
			break;
			
		case 3:
			[NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(twoTaps) object:nil];
			[self performSelector:@selector(threeTaps) withObject:nil afterDelay:.5];
			break;
			
		default:
			break;
	}
}

- (void)loadingFailed:(ISMSLoader*)theLoader withError:(NSError *)error
{
    //DDLogVerbose(@"[AsynchronousImageView] async cover art loading failed");
    
	[self.activityIndicator removeFromSuperview];
	self.activityIndicator = nil;
	
	self.coverArtDAO = nil;
	if ([self.delegate respondsToSelector:@selector(asyncImageViewLoadingFailed:withError:)])
	{
		[self.delegate asyncImageViewLoadingFailed:self withError:error];
	}
}

- (void)loadingFinished:(ISMSLoader*)theLoader
{
    //DDLogVerbose(@"[AsynchronousImageView] async cover art loading finished for: %@", self.coverArtId);
    
	[self.activityIndicator removeFromSuperview];
	self.activityIndicator = nil;
	
   //DLog(@"isLarge: %@", NSStringFromBOOL(self.isLarge));
   //DLog(@"delegate: %@", self.delegate);
	
	self.image = self.coverArtDAO.coverArtImage;
    self.coverArtDAO = nil;
	
	if ([self.delegate respondsToSelector:@selector(asyncImageViewFinishedLoading:)])
	{
		[self.delegate asyncImageViewFinishedLoading:self];
	}
}

@end

File: Classes/AsynchronousImageView.swift
//
//  AsynchronousImageView.swift
//  iSub
//
//  Created by Benjamin Baron on 6/16/16.
//  Copyright © 2016 Ben Baron. All rights reserved.
//

import UIKit

@objc protocol AsynchronousImageViewDelegate {
    func asyncImageViewFinishedLoading(_ asyncImageView: AsynchronousImageView)
    func asyncImageViewLoadingFailed(_ asyncImageView: AsynchronousImageView, error: Error)
}

class AsynchronousImageView: UIImageView, ISMSLoaderDelegate {
    weak var delegate: AsynchronousImageViewDelegate?
    
    var coverArtId: String? {
        didSet {
            reloadCoverArt()
        }
    }
    var coverArtDAO: SUSCoverArtDAO?
    var large = false
    
    fileprivate var activityIndicator: UIActivityIndicatorView?
    
    init() {
        super.init(frame: CGRect.zero)
    }
    
    init(frame: CGRect, coverArtId: String, large: Bool, delegate: AsynchronousImageViewDelegate) {
        super.init(frame: frame)
        
        self.coverArtId = coverArtId
        self.large = large
        self.delegate = delegate
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    fileprivate func reloadCoverArt() {
        // Make sure old activity indicator is gone
        activityIndicator?.removeFromSuperview()
        activityIndicator = nil;
        
        if let coverArtDAO = coverArtDAO {
            coverArtDAO.cancelLoad()
            coverArtDAO.delegate = nil
        }
        coverArtDAO = nil
        
        self.image = SUSCoverArtDAO.defaultCoverArtImage(forSize: large)
        
        if let coverArtId = coverArtId {
            coverArtDAO = SUSCoverArtDAO(delegate: self, coverArtId: coverArtId, isLarge: large)
            if coverArtDAO!.isCoverArtCached {
                self.image = coverArtDAO!.coverArtImage()
            } else if large {
                activityIndicator = UIActivityIndicatorView(activityIndicatorStyle: .whiteLarge)
                activityIndicator!.autoresizingMask = [.flexibleLeftMargin, .flexibleRightMargin, .flexibleTopMargin, .flexibleBottomMargin]
                activityIndicator!.center = CGPoint(x: self.width / 2.0, y: self.height / 2.0)
                self.addSubview(activityIndicator!)
                activityIndicator?.startAnimating()
            }
            coverArtDAO!.startLoad()
        }
    }
    
    //
    // MARK: - Loading Delegate -
    //
    
    func loadingFinished(_ theLoader: ISMSLoader) {
        activityIndicator?.removeFromSuperview()
        activityIndicator = nil
        
        image = coverArtDAO?.coverArtImage()
        coverArtDAO = nil
        
        if let delegate = delegate {
            delegate.asyncImageViewFinishedLoading(self)
        }
    }
    
    public func loadingFailed(_ theLoader: ISMSLoader, withError error: Error) {
        activityIndicator?.removeFromSuperview()
        activityIndicator = nil
        
        coverArtDAO = nil
        if let delegate = delegate {
            delegate.asyncImageViewLoadingFailed(self, error: error)
        }
    }
    
    //
    // MARK: - Touch Handling -
    //
    
    @objc fileprivate func oneTap() {
        
    }
    
    @objc fileprivate func twoTaps() {
        
    }
    
    @objc fileprivate func threeTaps() {
        
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        if let touch = touches.first {
            switch touch.tapCount {
            case 1:
                self.perform(#selector(AsynchronousImageView.oneTap), with: nil, afterDelay: 0.5)
            case 2:
                NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(AsynchronousImageView.oneTap), object: nil)
                self.perform(#selector(AsynchronousImageView.twoTaps), with: nil, afterDelay: 0.5)
            case 3:
                NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(AsynchronousImageView.twoTaps), object: nil)
                self.perform(#selector(AsynchronousImageView.threeTaps), with: nil, afterDelay: 0.5)
            default:
                break
            }
        }
    }
}
